### PR TITLE
perf: optimize array_remove for scalar needle

### PR DESCRIPTION
**Pull request by lyne7-sc**
## Which issue does this PR close?

- Closes #.

## Rationale for this change

Similar to #22387 (array_replace scalar optimization)

`array_remove` / `array_remove_n` / `array_remove_all` perform element-wise comparison by invoking `compare_element_to_list` against each row's sub-array individually. When the needle is a scalar, this can be optimized by performing a single vectorized `distinct` comparison over the entire flattened values buffer.

## What changes are included in this PR?

- Add a specialized removal kernel (`general_remove_with_scalar`) that uses `arrow_ord::cmp::distinct` with `Scalar` wrapper for a single bulk comparison pass over the flat values buffer.
- Extend SLT tests with multi-row scalar-argument coverage, NULL-containing arrays, empty-array edge cases, boundary `n` values, and LargeList type coverage.

### Benchmarks
```
group                                                                    main                                   optimized
-----                                                                    ----                                   ---------
array_remove_all_int64/remove/list size: 10, num_rows: 4000              4.35   856.8±97.81µs        ? ?/sec    1.00    196.9±4.48µs        ? ?/sec
array_remove_all_int64/remove/list size: 100, num_rows: 10000            1.90      5.5±0.09ms        ? ?/sec    1.00      2.9±0.09ms        ? ?/sec
array_remove_all_int64/remove/list size: 500, num_rows: 10000            1.35     19.2±0.21ms        ? ?/sec    1.00     14.2±0.48ms        ? ?/sec
array_remove_all_int64_nested/remove/list size: 10, num_rows: 4000       1.00      7.1±0.12ms        ? ?/sec    1.04      7.4±0.12ms        ? ?/sec
array_remove_all_int64_nested/remove/list size: 100, num_rows: 3000      1.00     36.5±0.39ms        ? ?/sec    1.05     38.3±2.61ms        ? ?/sec
array_remove_all_int64_nested/remove/list size: 300, num_rows: 1500      1.01     53.5±2.26ms        ? ?/sec    1.00     53.0±0.99ms        ? ?/sec
array_remove_boolean/remove/list size: 10, num_rows: 4000                3.83    813.9±7.08µs        ? ?/sec    1.00    212.4±2.28µs        ? ?/sec
array_remove_boolean/remove/list size: 100, num_rows: 10000              2.73      3.7±0.03ms        ? ?/sec    1.00  1364.7±177.83µs        ? ?/sec
array_remove_boolean/remove/list size: 500, num_rows: 10000              2.34      9.8±0.14ms        ? ?/sec    1.00      4.2±0.25ms        ? ?/sec
array_remove_fixed_size_binary/remove/list size: 10, num_rows: 4000      3.16   918.2±16.76µs        ? ?/sec    1.00    290.6±9.79µs        ? ?/sec
array_remove_fixed_size_binary/remove/list size: 100, num_rows: 10000    1.56      6.9±0.13ms        ? ?/sec    1.00      4.4±0.15ms        ? ?/sec
array_remove_fixed_size_binary/remove/list size: 500, num_rows: 10000    1.17     27.7±0.84ms        ? ?/sec    1.00     23.6±2.04ms        ? ?/sec
array_remove_int64/remove/list size: 10, num_rows: 4000                  4.55    825.7±6.30µs        ? ?/sec    1.00    181.3±4.32µs        ? ?/sec
array_remove_int64/remove/list size: 100, num_rows: 10000                3.35      3.8±0.11ms        ? ?/sec    1.00  1135.6±54.87µs        ? ?/sec
array_remove_int64/remove/list size: 500, num_rows: 10000                2.04     10.3±0.35ms        ? ?/sec    1.00      5.1±0.39ms        ? ?/sec
array_remove_int64_nested/remove/list size: 10, num_rows: 4000           1.00      7.1±0.18ms        ? ?/sec    1.02      7.2±0.07ms        ? ?/sec
array_remove_int64_nested/remove/list size: 100, num_rows: 3000          1.00     36.1±1.35ms        ? ?/sec    1.07     38.5±3.67ms        ? ?/sec
array_remove_int64_nested/remove/list size: 300, num_rows: 1500          1.00     51.7±0.57ms        ? ?/sec    1.05     54.1±2.13ms        ? ?/sec
array_remove_n_int64/remove/list size: 10, num_rows: 4000                4.43    845.3±5.00µs        ? ?/sec    1.00    190.6±2.84µs        ? ?/sec
array_remove_n_int64/remove/list size: 100, num_rows: 10000              2.29      4.7±0.11ms        ? ?/sec    1.00      2.0±0.12ms        ? ?/sec
array_remove_n_int64/remove/list size: 500, num_rows: 10000              1.63     14.8±0.42ms        ? ?/sec    1.00      9.0±0.51ms        ? ?/sec
array_remove_n_int64_nested/remove/list size: 10, num_rows: 4000         1.00      7.0±0.09ms        ? ?/sec    1.29      8.9±3.44ms        ? ?/sec
array_remove_n_int64_nested/remove/list size: 100, num_rows: 3000        1.00     36.6±0.42ms        ? ?/sec    1.03     37.7±0.68ms        ? ?/sec
array_remove_n_int64_nested/remove/list size: 300, num_rows: 1500        1.00     52.7±3.68ms        ? ?/sec    1.03     54.5±4.49ms        ? ?/sec
array_remove_strings/remove/list size: 10, num_rows: 4000                2.50  1144.6±21.95µs        ? ?/sec    1.00   457.0±14.15µs        ? ?/sec
array_remove_strings/remove/list size: 100, num_rows: 10000              1.42     10.5±1.16ms        ? ?/sec    1.00      7.4±0.34ms        ? ?/sec
array_remove_strings/remove/list size: 500, num_rows: 10000              1.12     39.8±0.91ms        ? ?/sec    1.00     35.5±1.51ms        ? ?/sec
```

## Are these changes tested?

Yes, existing and new SLT edge-case tests in `array_remove.slt`.

## Are there any user-facing changes?

No.

### CODE REVIEW

File: datafusion/functions-nested/src/remove.rs [hunk 1]
```diff
@@ -468,6 +531,98 @@ fn general_remove<OffsetSize: OffsetSizeTrait>(
...
+    let list_field = match list_array.data_type() {
+        DataType::List(field) | DataType::LargeList(field) => field,
+        _ => {
+            return exec_err!(
+                "Expected List or LargeList data type, got {:?}",
+                list_array.data_type()
+            );
+        }
+    };
+    let original_data = list_array.values().to_data();
```
Comment (neilconway) [inline datafusion/functions-nested/src/remove.rs]: This will be inefficient for sliced arrays.

File: datafusion/functions-nested/src/remove.rs [hunk 2]
```diff
@@ -357,6 +381,45 @@ fn array_remove_internal(
...
+    match array.data_type() {
+        DataType::List(_) => {
+            let list_array = array.as_list::<i32>();
+            general_remove_with_scalar::<i32>(list_array, needle, arr_n)
+        }
+        DataType::LargeList(_) => {
+            let list_array = array.as_list::<i64>();
+            general_remove_with_scalar::<i64>(list_array, needle, arr_n)
+        }
+        array_type => exec_err!("array_remove does not support type '{array_type}'."),
```
Comment (neilconway) [inline datafusion/functions-nested/src/remove.rs]: This is called by more than just `array_remove`; can we improve the error message?

File: datafusion/functions-nested/src/remove.rs [hunk 3]
```diff
@@ -468,6 +531,98 @@ fn general_remove<OffsetSize: OffsetSizeTrait>(
...
+                if let Some(bs) = pending_batch_to_retain {
+                    mutable.extend(0, start + bs, start + i);
+                    copied += i - bs;
+                    pending_batch_to_retain = None;
+                }
+                removed += 1;
+            } else if pending_batch_to_retain.is_none() {
+                pending_batch_to_retain = Some(i);
+            }
+        }
```
Comment (neilconway) [inline datafusion/functions-nested/src/remove.rs]: I wonder if it would be possible to iterate only over the "false" bits, e.g., by negating the buffer and looking at `BooleanBuffer::set_indices`.

File: datafusion/functions-nested/src/remove.rs [hunk 4]
```diff
@@ -468,6 +531,98 @@ fn general_remove<OffsetSize: OffsetSizeTrait>(
...
+            offsets.push(offsets[row_index] + OffsetSize::usize_as(end - start));
+            continue;
+        }
+
+        let max_removals = n.min(num_to_remove as i64);
+        let mut removed = 0i64;
+        let mut copied = 0usize;
+        let mut pending_batch_to_retain: Option<usize> = None;
+        for (i, keep) in eq_array.iter().enumerate() {
+            if keep == Some(false) && removed < max_removals {
```
Comment (neilconway) [inline datafusion/functions-nested/src/remove.rs]: Can we break from the loop once we hit `max_removals`?

**Review by neilconway — COMMENTED** [review]
Nice performance win!

File: datafusion/functions-nested/src/remove.rs (hunk 2 above)
Comment (lyne7-sc) [reply datafusion/functions-nested/src/remove.rs]: sure, updated.

File: datafusion/functions-nested/src/remove.rs (hunk 3 above)
Comment (lyne7-sc) [reply datafusion/functions-nested/src/remove.rs]: Great suggestion. Benchmarks show a ~20–40% improvement with this optimization.

File: datafusion/functions-nested/src/remove.rs (hunk 4 above)
Comment (lyne7-sc) [reply datafusion/functions-nested/src/remove.rs]: Good point. now break early once `max_removals` is reached.

File: datafusion/functions-nested/src/remove.rs (hunk 3 above)
Comment (neilconway) [reply datafusion/functions-nested/src/remove.rs]: Amazing!

File: datafusion/functions-nested/src/remove.rs (hunk 1 above)
Comment (lyne7-sc) [reply datafusion/functions-nested/src/remove.rs]: I now slice the values to the range actually referenced by the offsets.

That said, I wanted to understand your concern better: when a `GenericListArray` is sliced, `values()` returns the full underlying array, and `to_data()` on it wraps the existing buffer references into `ArrayData` without copying. So the main downside I could identify is that `Capacities::Array(original_data.len())` over-estimates the pre-allocation for sliced inputs. Were you thinking of a different inefficiency, or is the over-allocation what you had in mind?

File: datafusion/functions-nested/src/remove.rs [hunk 5]
```diff
@@ -468,6 +533,113 @@ fn general_remove<OffsetSize: OffsetSizeTrait>(
...
+        if num_to_remove == 0 {
+            mutable.extend(0, start, end);
+            offsets.push(offsets[row_index] + OffsetSize::usize_as(row_len));
+            continue;
+        }
+
+        let max_removals = n.min(num_to_remove as i64) as usize;
+
+        // Iterate only over the positions that need removal using set_indices,
+        // which is more efficient than scanning every bit.
```
Comment (neilconway) [inline datafusion/functions-nested/src/remove.rs]: Might be worth elaborating that the win here is mostly because we expect the # of values-to-remove is a lot smaller than the total array size, which it usually (but not always) will be.

File: datafusion/functions-nested/src/remove.rs [hunk 6]
```diff
@@ -468,6 +533,113 @@ fn general_remove<OffsetSize: OffsetSizeTrait>(
...
+    offsets.push(OffsetSize::zero());
+
+    let mut mutable = MutableArrayData::with_capacities(
+        vec![&original_data],
+        false,
+        Capacities::Array(original_data.len()),
+    );
+    let nulls = list_array.nulls().cloned();
+    let keep_mask =
+        arrow_ord::cmp::distinct(list_array.values(), &Scalar::new(Arc::clone(needle)))?;
```
Comment (neilconway) [inline datafusion/functions-nested/src/remove.rs]: This will call the distinct kernel on all the elements in the value buffer, not just the ones that are visible in a sliced array.

File: datafusion/functions-nested/src/remove.rs (hunk 1 above)
Comment (neilconway) [reply datafusion/functions-nested/src/remove.rs]: The over-allocation was one part, but the bigger concern is calling the `distinct` kernel on the entire values buffer (see other comment).

**Review by neilconway — APPROVED** [review]
Awesome, nice work on this! lgtm. Maybe update the benchmark numbers in the PR description if you get a chance.

**Comment by lyne7-sc** [comment]
Thanks @neilconway! Really appreciate the thorough and speedy review 🙏 Great suggestion — running benchmarks now and will update the pr description with fresh numbers shortly.

File: datafusion/functions-nested/src/remove.rs [hunk 7]
```diff
@@ -357,6 +381,47 @@ fn array_remove_internal(
...
+            "array_remove/array_remove_n/array_remove_all does not support type '{array_type}'."
+        ),
+    }
+}
+
+/// Removes elements matching a scalar needle from a list array.
+///
+/// Uses a bulk `distinct` comparison for non-null, non-nested scalar elements,
+/// falling back to the per-row `general_remove` path for null or nested types.
+fn remove_with_scalar_needle(
```
Comment (Jefffrey) [inline datafusion/functions-nested/src/remove.rs]: I find this function a little confusing, since the code paths can then be like:

```
invoke with args -> remove_with_scalar_needle -> array_remove_internal (fallback)
```

Perhaps we can hoist this null/nested check earlier to remove need for this function? Or the benefits of centralizing the check outweighs it?

File: datafusion/functions-nested/src/remove.rs [hunk 8]
```diff
@@ -214,7 +229,23 @@ impl ScalarUDFImpl for ArrayRemoveN {
     }
 
     fn invoke_with_args(&self, args: ScalarFunctionArgs) -> Result<ColumnarValue> {
-        make_scalar_function(array_remove_n_inner)(&args.args)
+        let [list_arg, element_arg, max_arg] =
+            take_function_args(self.name(), &args.args)?;
+        let num_rows = args.number_rows;
+        let list_array = list_arg.to_array(num_rows)?;
+        let max_array = max_arg.to_array(num_rows)?;
+        let arr_n = as_int64_array(&max_array)?.values().to_vec();
```
Comment (Jefffrey) [inline datafusion/functions-nested/src/remove.rs]: We're ignoring nulls in `max_arg` here, though I think this may be an existing issue

File: datafusion/functions-nested/src/remove.rs [hunk 9]
```diff
@@ -468,6 +533,114 @@ fn general_remove<OffsetSize: OffsetSizeTrait>(
...
+        }
+    };
+
+    let list_offsets = list_array.offsets();
+    let first_offset = list_offsets[0].to_usize().unwrap();
+    let last_offset = list_offsets[list_offsets.len() - 1].to_usize().unwrap();
+    let values_range_len = last_offset - first_offset;
+    let values_slice = list_array.values().slice(first_offset, values_range_len);
+    let original_data = values_slice.to_data();
+    let mut offsets = Vec::<OffsetSize>::with_capacity(list_array.len() + 1);
```
Comment (Jefffrey) [inline datafusion/functions-nested/src/remove.rs]: Use [`OffsetBufferBuilder`](https://docs.rs/arrow/latest/arrow/array/struct.OffsetBufferBuilder.html) here

File: datafusion/functions-nested/src/remove.rs [hunk 10]
```diff
@@ -468,6 +533,114 @@ fn general_remove<OffsetSize: OffsetSizeTrait>(
...
+    offsets.push(OffsetSize::zero());
+
+    let mut mutable = MutableArrayData::with_capacities(
```
Comment (Jefffrey) [inline datafusion/functions-nested/src/remove.rs:577]: I wonder if an approach using `take` kernel could provide even more performance gains?

File: datafusion/functions-nested/src/remove.rs [hunk 11]
```diff
@@ -468,6 +533,114 @@ fn general_remove<OffsetSize: OffsetSizeTrait>(
...
+        vec![&original_data],
+        false,
+        Capacities::Array(original_data.len()),
+    );
+    let nulls = list_array.nulls().cloned();
+    let keep_mask =
+        arrow_ord::cmp::distinct(&values_slice, &Scalar::new(Arc::clone(needle)))?;
+    let remove_bits = match keep_mask.nulls() {
```
Comment (Jefffrey) [inline datafusion/functions-nested/src/remove.rs]: > `distinct` is similar to `neq`, only differing in null handling. In particular, two
operands are considered DISTINCT if they have a different value or if one of them is NULL
and the other isn’t. The result of `distinct` is never NULL.

https://docs.rs/arrow/latest/arrow/compute/kernels/cmp/fn.distinct.html

I don't think we should have handling around the null buffer of `keep_mask` according to the documentation

File: datafusion/functions-nested/src/remove.rs [hunk 12]
```diff
@@ -468,6 +533,114 @@ fn general_remove<OffsetSize: OffsetSizeTrait>(
...
 
+/// For each element of `list_array[i]`, removed up to `arr_n[i]` occurrences
+/// of `needle[0]` (scalar element broadcasted).
+///
+/// This is a specialized version of `general_remove` for scalar elements that
+/// uses bulk comparison for better performance.
+fn general_remove_with_scalar<OffsetSize: OffsetSizeTrait>(
+    list_array: &GenericListArray<OffsetSize>,
+    needle: &ArrayRef,
+    arr_n: &[i64],
```
Comment (Jefffrey) [inline datafusion/functions-nested/src/remove.rs]: Similar question to that of array_replace PR, in if we should just handle when needle & max are scalars only

File: datafusion/functions-nested/src/remove.rs (hunk 8 above)
Comment (lyne7-sc) [reply datafusion/functions-nested/src/remove.rs]: Fixed now, add `is_null(i)` is checked to substitute 0 for null positions.

File: datafusion/functions-nested/src/remove.rs (hunk 7 above)
Comment (lyne7-sc) [reply datafusion/functions-nested/src/remove.rs]: fixed, hoisted the scalar check into `invoke_with_args` directly

File: datafusion/functions-nested/src/remove.rs (hunk 10 above)
Comment (lyne7-sc) [reply datafusion/functions-nested/src/remove.rs:577]: I benchmarked both `take` and `filter` kernel approaches against the current `MutableArrayData::extend` path. Overall, `MutableArrayData` performs best — on small lists (size=10) take is faster (~10%), possibly by avoiding `MutableArrayData` initialization overhead, but on medium/large lists (size≥100) `MutableArrayData` tends to win decisively (take is 60–170% slower depending on type). For variable-length types (strings), the gap appears to widen further. 

One possible explanation is that take performs per-index random access for each element, whereas `MutableArrayData` may instead execute contiguous memcpy operations over memory regions?

Comment (lyne7-sc) [reply datafusion/functions-nested/src/remove.rs:577]: But, the current benchmarks use 10% needle density (sparse removals / high retention), which is likely the most common case? For dense removal workloads the trade-offs may shift, `take` or `filter` could become competitive when there are fewer contiguous ranges to memcpy.

Comment (Jefffrey) [reply datafusion/functions-nested/src/remove.rs:577]: That makes sense, especially as we're operating on list arrays anyway

**Comment by Jefffrey** [comment]
run benchmark array_remove

File: datafusion/functions-nested/src/remove.rs [hunk 13]
```diff
@@ -468,6 +542,105 @@ fn general_remove<OffsetSize: OffsetSizeTrait>(
...
+
+    let mut mutable = MutableArrayData::with_capacities(
+        vec![&original_data],
+        false,
+        Capacities::Array(original_data.len()),
+    );
+    let nulls = list_array.nulls().cloned();
+    let needle = scalar_needle.to_array_of_size(1)?;
+    let keep_mask = arrow_ord::cmp::distinct(&values_slice, &Scalar::new(needle))?;
+    let remove_bits = !keep_mask.values();
```
Comment (Jefffrey) [inline datafusion/functions-nested/src/remove.rs]: nit: we could use `not_distinct` instead of negating the `distinct` result

**Review by Jefffrey — APPROVED** [review]

**Comment by lyne7-sc** [comment]
@Jefffrey Addressed all comments. Thank you for the detailed review!

**Comment by Jefffrey** [comment]
Thanks @lyne7-sc & @neilconway
